### PR TITLE
[IPU] Support manually instantiating the `poptorch.DataLoader`

### LINKED ISSUE #10695: Support creating poptorch.DataLoader from within a LightningDataModule
## 🚀 Feature

Support creating `poptorch.DataLoader` from within a `LightningDataModule`

### Motivation

Related issue: #10694 

A user with knowledge of the IPU and the poptorch API cannot currently create a `poptorch.DataLoader` within their `LightningDataModule`.  This is limiting as there are a number of options that the `poptorch.DataLoader` provides that are not supported on the native PyTorch `DataLoader`.  Without support for this feature it is difficult for a user to manage advanced batching scenarios for the IPU.

### Pitch
Here is a minimized example showing a `BoringDataModule` that directly instantiates the `poptorch.DataLoader`

```python
class BoringDataModule(LightningDataModule):
    def get_options(self, training=False):
        plugin = self.trainer.accelerator.training_type_plugin

        if training:
            return plugin.training_opts

        return plugin.inference_opts

    def train_dataloader(self):
        opts = self.get_options(training=True)
        return poptorch.DataLoader(options=opts,
                                   dataset=RandomDataset(32, 64),
                                   batch_size=2)

    def val_dataloader(self):
        opts = self.get_options(training=False)
        return poptorch.DataLoader(options=opts,
                                   dataset=RandomDataset(32, 64),
                                   batch_size=2)

    def test_dataloader(self):
        opts = self.get_options(training=False)
        return poptorch.DataLoader(options=opts,
                                   dataset=RandomDataset(32, 64),
                                   batch_size=2)
```

### Alternatives

1. Don't use PyTorch Lightning with IPUs 😭 
2. Reimplement all advanced batching and data access options in Lightning from poptorch. This feels like obvious an anti-pattern.

cc @borda @SeanNaren

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes #10695

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

cc @borda @justusschock @awaelchli @ninginthecloud

### CODE REVIEW

File: pytorch_lightning/strategies/ipu.py [hunk 1]
```diff
@@ -214,10 +214,11 @@ def lightning_module(self) -> Optional["pl.LightningModule"]:
     def _convert_to_poptorch_loader(
         self, dataloader: DataLoader, sampler, mode: Optional[RunningStage] = None
     ) -> "poptorch.DataLoader":
-        dl_kwargs = _get_dataloader_init_kwargs(dataloader, sampler)
-        # Override to drop last uneven batch, as IPUs does not support uneven inputs.
-        dl_kwargs["drop_last"] = True
```
Comment (carmocca) [inline pytorch_lightning/strategies/ipu.py:219]: I'm unsure why we were dropping this key, they do support it:

https://docs.graphcore.ai/projects/poptorch-user-guide/en/latest/reference.html#poptorch.DataLoader

And drop_last does not mean uneven inputs, but whether we drop the uneven ones or pad with repeated samples

File: pytorch_lightning/utilities/data.py [hunk 2]
```diff
@@ -329,7 +329,13 @@ def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
         params = dict(inspect.signature(obj.__init__).parameters)
         params.pop("args", None)
         params.pop("kwargs", None)
+        cls = type(obj)
         for arg_name, arg_value in chain(zip(params, args), kwargs.items()):
+            if hasattr(cls, arg_name) and getattr(cls, arg_name).fset is None:
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:334]: This is unreleased code so does not need a CHANGELOG fix entry

File: pytorch_lightning/utilities/data.py [hunk 3]
```diff
@@ -329,7 +329,13 @@ def wrapper(obj: DataLoader, *args: Any, **kwargs: Any) -> None:
...
+                # the class defines a read-only (no setter) property of this name. it's likely that the implementation
+                # will set `self._arg_name = arg_value` in `__init__` which is the attribute returned by the `arg_name`
+                # property so we are fine skipping in that case
```
Comment (carmocca) [inline pytorch_lightning/utilities/data.py:337]: you know you got really confused about a bug when you write a 3 line comment for a 2 line change

File: tests/accelerators/test_ipu.py [hunk 4]
```diff
@@ -344,6 +344,38 @@ def test_autoreport(tmpdir):
     assert os.path.isfile(autoreport_path + "profile.pop")
 
 
+@RunIf(ipu=True)
+def test_manual_poptorch_dataloader(tmpdir):
+    model_options = poptorch.Options()
```
Comment (krshrimali) [inline tests/accelerators/test_ipu.py:349]: Uhmm, so we assume that `poptorch` will always be installed? Should we also make sure that it's available in the decorator `@RunIf(ipu=True)`?

**Review by krshrimali — APPROVED** [review]
Minor comments, but shouldn't block merging this PR. LGTM 🚀  (also, nice inline comments)

File: tests/accelerators/test_ipu.py (hunk 4 above)
Comment (carmocca) [reply tests/accelerators/test_ipu.py:349]: `poptorch` will be installed if `ipu=True`

Comment (krshrimali) [reply tests/accelerators/test_ipu.py:349]: Cool!

**Review by ananthsub — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
